cmd/cli/install.go

```diff
@@ -243,50 +154,17 @@ func (i *installCmd) loadOSMChart() error {
 func (i *installCmd) resolveValues() (map[string]interface{}, error) {
 	finalValues := map[string]interface{}{}
 
-	for _, val := range i.setOptions {
-		// parses Helm strvals line and merges into a map for the final overrides for values.yaml
-		if err := strvals.ParseInto(val, finalValues); err != nil {
-			return nil, errors.Wrap(err, "invalid format for --set")
-		}
+	if err := parseVal(i.setOptions, finalValues); err != nil {
+		return nil, errors.Wrap(err, "invalid format for --set")
 	}
 
 	valuesConfig := []string{
-		fmt.Sprintf("OpenServiceMesh.image.registry=%s", i.containerRegistry),
-		fmt.Sprintf("OpenServiceMesh.image.tag=%s", i.osmImageTag),
-		fmt.Sprintf("OpenServiceMesh.image.pullPolicy=%s", i.osmImagePullPolicy),
-		fmt.Sprintf("OpenServiceMesh.certificateManager=%s", i.certificateManager),
-		fmt.Sprintf("OpenServiceMesh.vault.host=%s", i.vaultHost),
-		fmt.Sprintf("OpenServiceMesh.vault.protocol=%s", i.vaultProtocol),
-		fmt.Sprintf("OpenServiceMesh.vault.token=%s", i.vaultToken),
-		fmt.Sprintf("OpenServiceMesh.vault.role=%s", i.vaultRole),
-		fmt.Sprintf("OpenServiceMesh.certmanager.issuerName=%s", i.certManagerIssuerName),
-		fmt.Sprintf("OpenServiceMesh.certmanager.issuerKind=%s", i.certManagerIssuerKind),
-		fmt.Sprintf("OpenServiceMesh.certmanager.issuerGroup=%s", i.certManagerIssuerGroup),
-		fmt.Sprintf("OpenServiceMesh.serviceCertValidityDuration=%s", i.serviceCertValidityDuration),
-		fmt.Sprintf("OpenServiceMesh.prometheus.retention.time=%s", i.prometheusRetentionTime),
-		fmt.Sprintf("OpenServiceMesh.enableDebugServer=%t", i.enableDebugServer),
-		fmt.Sprintf("OpenServiceMesh.enablePermissiveTrafficPolicy=%t", i.enablePermissiveTrafficPolicy),
-		fmt.Sprintf("OpenServiceMesh.deployPrometheus=%t", i.deployPrometheus),
-		fmt.Sprintf("OpenServiceMesh.enablePrometheusScraping=%t", i.enablePrometheusScraping),
-		fmt.Sprintf("OpenServiceMesh.deployGrafana=%t", i.deployGrafana),
-		fmt.Sprintf("OpenServiceMesh.enableFluentbit=%t", i.enableFluentbit),
 		fmt.Sprintf("OpenServiceMesh.meshName=%s", i.meshName),
-		fmt.Sprintf("OpenServiceMesh.enableEgress=%t", i.enableEgress),
-		fmt.Sprintf("OpenServiceMesh.deployJaeger=%t", i.deployJaeger),
-		fmt.Sprintf("OpenServiceMesh.envoyLogLevel=%s", strings.ToLower(i.envoyLogLevel)),
-		fmt.Sprintf("OpenServiceMesh.controllerLogLevel=%s", strings.ToLower(i.controllerLogLevel)),
 		fmt.Sprintf("OpenServiceMesh.enforceSingleMesh=%t", i.enforceSingleMesh),
 	}
 
-	if i.containerRegistrySecret != "" {
-		valuesConfig = append(valuesConfig, fmt.Sprintf("OpenServiceMesh.imagePullSecrets[0].name=%s", i.containerRegistrySecret))
-	}
-
-	for _, val := range valuesConfig {
-		// parses Helm strvals line and merges into a map for the final overrides for values.yaml
-		if err := strvals.ParseInto(val, finalValues); err != nil {
-			return nil, err
-		}
+	if err := parseVal(valuesConfig, finalValues); err != nil {
```

Comment:
> The changes to `resolveValues()` do introduce a slight change in behavior. Before, all the values that could be set by a flag were "overridden" in the Helm release, whether or not the flag for a value was explicitly set. This change will only keep that behavior for `OpenServiceMesh.meshName` and `OpenServiceMesh.enforceSingleMesh`. All other values will only be stored as overridden values for the release if it was passed to the command line with `--set`.
> 
Correct, but I believe this is an improvement that eliminates the need for default values to be duplicated in two places(`install.go` and `values.yml`).
> This shouldn't impact install at all, but does force us to consider how we're handling them for upgrade. [94b3ffa](https://github.com/openservicemesh/osm/commit/94b3ffa8fd9284039045a6d3dc31c5784846863e) describes the precedence order of values as:
> 
> 1. New chart defaults, overridden by
> 2. Old chart defaults, overridden by
> 3. Old release overrides, overridden by
> 4. New release overrides
> 
> Since `osm mesh upgrade` already only sets values when they've explicitly been set on the command line, these changes shouldn't affect its existing behavior.
> 
> Overall, the Helm release will look a bit different, but users probably won't see any changes in behavior. We could emulate the old behavior while still using the rest of these changes if we find it's an issue.

Sorry if I misunderstood, but I don't think this change would affect `mesh upgrade` or its behavior. Since the update is only removing explicit override which is anyways overridden by chart's `values.yaml`.